### PR TITLE
Fix TPL usage in Amesos2 (#7536)

### PR DESCRIPTION
**Pull request by bartlettroscoe**
This now is closer to following instructions outlined at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

I can't test all of this because I don't have all of these TPLs installed and
there is no PR tseting for some of them like MUMPS.

This was motivated by #7536.

## How was this tested?

On my SNL RHEL6 machine using ATDM and SEMS, I first set up a configure directory with:

```
$ cd ~/Trilinos.base/BUILDS/ATDM/SEMS-RHEL6/CHECKIN/

$ ./checkin-test-atdm.sh gnu-openmp-dbg-pt --enable-packages=Amesos2 --configure

$ cd gnu-openmp-dbg-pt

$ . load-env.sh 
Hostname 'crf450.srn.sandia.gov' matches known ATDM host 'crf450.srn.sandia.gov' and system 'sems-rhel6'
Setting compiler and build options for build-name 'gnu-openmp-dbg-pt'
Using SEMS RHEL6 compiler stack GNU-7.2.0 to build DEBUG code with Kokkos node type OPENMP
```

Then I tested the enabling just `-DTrilinos_ENABLE_Amesos2=ON -DTPL_ENABLE_MUMPS=ON` as @modkin in #7536 with:

```
$  rm -r CMake* *.out 

$ ./do-configure.base -DTrilinos_ENABLE_Amesos2=ON -DTPL_ENABLE_MUMPS=ON \
  -DTPL_MUMPS_INCLUDE_DIRS=/usr/include -DTPL_MUMPS_LIBRARIES=/usr/lib64/libblas.so \
  &> configure.out

$ grep MUMPS configure.out 
Explicitly enabled TPLs on input (by user):  MPI BLAS LAPACK Boost HDF5 Netcdf MUMPS BoostLib DLlib 9
-- Setting Amesos2_ENABLE_MUMPS=ON since TPL_ENABLE_MUMPS=ON
Final set of enabled TPLs:  MPI BLAS LAPACK Boost HDF5 Netcdf MUMPS BoostLib DLlib 9
Processing enabled TPL: MUMPS (enabled explicitly, disable with -DTPL_ENABLE_MUMPS=OFF)
-- MUMPS_LIBRARY_NAMES='dmumps;libmumps_common;pord'
-- TPL_MUMPS_LIBRARIES='/usr/lib64/libblas.so'
-- TPL_MUMPS_INCLUDE_DIRS='/usr/include'
  *****Amesos2 will not provide MUMPS Support with Tpetra_INST_INT_INT

$ grep MUMPS packages/amesos2/src/Amesos2_config.h 
/* Defined if you want to enable amesos2-MUMPS */
/* #undef HAVE_AMESOS2_MUMPS */
```

Then I configured with `-DTrilinos_ENABLE_Amesos2=ON -DTpetra_INST_INT_INT=ON -DTPL_ENABLE_MUMPS=ON ` with:

```
$ rm -r CMake* *.out

$ ./do-configure.base -DTrilinos_ENABLE_Amesos2=ON -DTpetra_INST_INT_INT=ON \
  -DTPL_ENABLE_MUMPS=ON \
  -DTPL_MUMPS_INCLUDE_DIRS=/usr/include -DTPL_MUMPS_LIBRARIES=/usr/lib64/libblas.so \
  &> configure.out

$ grep MUMPS configure.out
 MPI BLAS LAPACK Boost HDF5 Netcdf MUMPS BoostLib DLlib 9
-- Setting Amesos2_ENABLE_MUMPS=ON since TPL_ENABLE_MUMPS=ON
Final set of enabled TPLs:  MPI BLAS LAPACK Boost HDF5 Netcdf MUMPS BoostLib DLlib 9
Processing enabled TPL: MUMPS (enabled explicitly, disable with -DTPL_ENABLE_MUMPS=OFF)
-- MUMPS_LIBRARY_NAMES='dmumps;libmumps_common;pord'
-- TPL_MUMPS_LIBRARIES='/usr/lib64/libblas.so'
-- TPL_MUMPS_INCLUDE_DIRS='/usr/include'
-- Amesos2_MUMPS_UnitTests_MPI_4: Added test (BASIC, NUM_MPI_PROCS=4, PROCESSORS=4)!

$ grep MUMPS packages/amesos2/src/Amesos2_config.h 
/* Defined if you want to enable amesos2-MUMPS */
#define HAVE_AMESOS2_MUMPS
```

That looks correct to me.

----

NOTE: Setting `TPL_MUMPS_INCLUDE_DIRS` and `TPL_MUMPS_LIBRARIES` to some dummy value bypasses checking and allows you to complete the configure stage even if you don't have MUMPS installed on your system (which I don't).

### CODE REVIEW

**Comment by ndellingwood** [comment]
@MicheldeMessieres when you have a moment can you test changes in this PR with your configurations for enabling the triangular solvers using cholmod, cusparse, with metis (where relevant)? There is some cleanup and fixes for mumps, but it may remove configuration options that need to be explicitly set for some of the triangular solver capabilities you have worked on recently.

**Comment by bartlettroscoe** [comment]
Why was this never approved?

**Comment by bartlettroscoe** [comment]
@srajama1 (Trilinos Linear Solver Product Lead), @trilinos/amesos2,

FYI: tI was confirmed by @modkin in https://github.com/trilinos/Trilinos/issues/7536#issuecomment-647346110 that this PR fixes the problem reported in #7536.  If I fix the merge conflicts, can we please merge this to the Trilinos 'develop' branch?

**Comment by ndellingwood** [comment]
Hi @bartlettroscoe I must've missed the previous amesos2 ping on this, sorry for the delay, will take a look now

**Review by ndellingwood — COMMENTED** [review]
Thanks @bartlettroscoe for contributing the PR, a question I have for the common pattern of changes:
There is code in amesos2 guarded by `HAVE_AMESOS2_{TPL}, TPL in {Cholmod, MUMPS, CUSPARSE, CUBLAS}`, will this removal code break those code paths? None of these solvers are enabled in PR testing afaik

**Comment by bartlettroscoe** [comment]
> 

No.  As described in:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

when your package lists a TPL as an optional TPL (i.e. in the list `LIB_OPTIONAL_TPLS ... <newTpl> ...` in  `Dependencies.cmake` file), TriBITS automatically defines the CMake cache var `<Package>_ENABLE_<newTpl>` and makes sure it is set correctly and defines the non-cache var `HAVE_<PACKAGE_NAME_UC>_<OPTIONAL_DEP_TPL_NAME_UC>` to get picked up when you configure your `<Package>_config.h` file.  All you have to do manually is add an entry to `<Package>_config.h` and use that macro in your C++ code and tests.  And in CMake logic, you key off of `IF (${PACKAGE_NAME}_ENABLE_${OPTIONAL_DEP_TPL_NAME})`.

Please let me know if the description of this at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

is not sufficient.

I will add some more comments to the patch in this PR to make this more clear.

File: packages/amesos2/CMakeLists.txt [hunk 1]
```diff
@@ -61,11 +62,6 @@ TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_TIMERS
 #
 
 
-IF(${PACKAGE_NAME}_ENABLE_Cholmod)
-  SET(HAVE_AMESOS2_CHOLMOD ON )
```
Comment (bartlettroscoe) [inline packages/amesos2/CMakeLists.txt:65]: No need for this.  TriBITS automatically defines and sets this var to the correct value.  See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

File: packages/amesos2/CMakeLists.txt [hunk 2]
```diff
@@ -14,6 +14,7 @@ SET(${PACKAGE_NAME_UC}_RELEASE_DATE "07/28/2011")
 
 IF (${PACKAGE_NAME}_ENABLE_MUMPS AND NOT Tpetra_INST_INT_INT)
    MESSAGE(WARNING "*****Amesos2 will not provide MUMPS Support with Tpetra_INST_INT_INT OFF.*****")
+  SET(HAVE_${PACKAGE_NAME_UC}_MUMPS OFF)
```
Comment (bartlettroscoe) [inline packages/amesos2/CMakeLists.txt:17]: NOTE: This special logic that is not supported directly by TriBITS but it seems to be what is desired.

File: packages/amesos2/CMakeLists.txt [hunk 3]
```diff
@@ -107,36 +103,6 @@ TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_ShyLU_NodeBasker
   ${Amesos2_HAVE_SHYLUBASKER}
 )
 
-TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_MUMPS
-  HAVE_AMESOS2_MUMPS
-  "Enable MUMPS in Amesos2"
-  OFF
-)
```
Comment (bartlettroscoe) [inline packages/amesos2/CMakeLists.txt:114]: No need for for this. TriBITS already defines these vars automatically. See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

File: packages/amesos2/CMakeLists.txt [hunk 4]
```diff
@@ -107,36 +103,6 @@ TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_ShyLU_NodeBasker
...
-
-ASSERT_DEFINED(TPL_ENABLE_METIS)
-TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_METIS
-  HAVE_AMESOS2_METIS
-  "Enable METIS in Amesos2"
-  "${TPL_ENABLE_METIS}"
-  )
-ASSERT_DEFINED(${PACKAGE_NAME}_ENABLE_METIS)
```
Comment (bartlettroscoe) [inline packages/amesos2/CMakeLists.txt:122]: No need for for this. TriBITS already defines these vars automatically. See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

File: packages/amesos2/CMakeLists.txt [hunk 5]
```diff
@@ -107,36 +103,6 @@ TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_ShyLU_NodeBasker
...
-
-ASSERT_DEFINED(TPL_ENABLE_CUSPARSE)
-TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_CUSPARSE
-  HAVE_AMESOS2_CUSPARSE
-  "Enable cuSPARSE in Amesos2"
-  "${TPL_ENABLE_CUSPARSE}"
-  )
-ASSERT_DEFINED(${PACKAGE_NAME}_ENABLE_CUSPARSE)
```
Comment (bartlettroscoe) [inline packages/amesos2/CMakeLists.txt:130]: No need for for this. TriBITS already defines these vars automatically. See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

File: packages/amesos2/CMakeLists.txt [hunk 6]
```diff
@@ -107,36 +103,6 @@ TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_ShyLU_NodeBasker
...
-
-ASSERT_DEFINED(TPL_ENABLE_CUSOLVER)
-TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_CUSOLVER
-  HAVE_AMESOS2_CUSOLVER
-  "Enable cuSOLVER in Amesos2"
-  "${TPL_ENABLE_CUSOLVER}"
-  )
-ASSERT_DEFINED(${PACKAGE_NAME}_ENABLE_CUSOLVER)
```
Comment (bartlettroscoe) [inline packages/amesos2/CMakeLists.txt:138]: No need for for this. TriBITS already defines these vars automatically. See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

File: packages/amesos2/cmake/Dependencies.cmake [hunk 7]
```diff
@@ -8,6 +8,6 @@ SET(TEST_REQUIRED_DEP_PACKAGES)
 SET(TEST_OPTIONAL_DEP_PACKAGES ShyLU_NodeBasker ShyLU_NodeTacho Kokkos TrilinosSS)
 # SET(LIB_REQUIRED_DEP_TPLS SuperLU)
 SET(LIB_REQUIRED_DEP_TPLS )
-SET(LIB_OPTIONAL_DEP_TPLS MPI SuperLU SuperLUMT SuperLUDist UMFPACK PARDISO_MKL ParMETIS METIS Cholmod MUMPS)
+SET(LIB_OPTIONAL_DEP_TPLS MPI SuperLU SuperLUMT SuperLUDist UMFPACK PARDISO_MKL ParMETIS METIS Cholmod MUMPS CUSPARSE CUSOLVER)
```
Comment (bartlettroscoe) [inline packages/amesos2/cmake/Dependencies.cmake:11]: This is the key.  You just need to tell TriBITS that `CUSPARSE` and `CUSOLVER` are optional TPLs for the `Amesos2` package and TriBITS takes care of defining and setting the following vars before processing the file `packages/amesos2/CMakeLists.txt`:

* `Amesos2_ENABLE_CUSPARSE`
* `HAVE_AMESOS2_CUSPARSE`
* `Amesos2_ENABLE_CUSOLVER`
* `HAVE_AMESOS2_CUSOLVER`

and likewise for all of these other optional TPLs.

**Review by bartlettroscoe — COMMENTED** [review]
Just adding some comments to try to help explain the changes here to be consistent with:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

**Review by ndellingwood — APPROVED** [review]
Thanks Ross for the explanations and the documentation link, PR and changes look good.

**Comment by ndellingwood** [comment]
@bartlettroscoe I'm going to remove the STALE label and add AUTOMERGE, though we may need to monitor for retest (OHPC and SRN systems are down today for equipment updates)

**Comment by ndellingwood** [comment]
Oop, you beat me to the label updates

**Pull request merged**